zebra-state/src/in_memory/chains.rs

```diff
@@ -0,0 +1,37 @@
+use super::Error;
+use crate::{Request, Response};
+use futures::prelude::*;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::Service;
+
+/// A service wrapper that tracks multiple chains, handles reorgs, and persists
+/// blocks to disk once they're past the reorg limit
+pub(crate) struct ChainsState<S> {
```

Comment:
My plan is to do this via composition of multiple services, rather than this being a book keeping struct that is owned by each service. This may not work out in the end but I'm optimistic.